sdk/python/kfp/components/pipeline_task.py

```diff
@@ -570,3 +468,56 @@ def my_pipeline():
                 )
             self._task_spec.dependent_tasks.append(task.name)
         return self
+
+
+# TODO: this function should ideally be in the function kfp.components.structures.check_placeholder_references_valid_io_name, which does something similar, but this causes the exception to be raised at component definition time, rather than compile time. This would break tests that load v1 component YAML, even though that YAML is invalid.
```

Comment:
> but this causes the exception to be raised at component definition time, rather than compile time. This would break tests that load v1 component YAML, even though that YAML is invalid.

Should we change the test to assertRaises on component definition then?

Reply:
I think it's possible my comment is unclear. Let me try again and let me know if this helps:

The validation logic (not this exact function) lived in  `PipelineTask.__init__` before this PR. However, while working on this placeholder refactoring, I noticed that the validation can be fully completed using only the component definition (it does not require knowing what the component arguments are). It therefore ought to live `component_factory.create_component_from_func`.

While working on this PR, I attempted to move it there, but this ended up breaking the `kubeflow-pipelines-component-yaml` test suite, since it caught errored(?) v1 component YAML on load, rather than instantiation as a task.

> Should we change the test to assertRaises on component definition then?

To avoid expanding the scope of this PR, I moved the logic (albeit refactored for clarity) back to its original location `PipelineTask.__init__`, maintaining the current state. So, the `assertRaises` still needs to be on the instantiation. I left a `TODO` to indicate that this should be dealt with, as I am opting not to within this PR.

Does that make sense? Can you let me know if you have any suggestions for how the comment can be clearer?

Reply:
Make sense, and it's fine by me to keep it here for now.